File: EmarsysMobileShop/EmarsysMobileShop/RecommendedItemsCell.m
//
//  RecommendedItemsCell.m
//  EmarsysMobileShop
//

#import "RecommendedItemsCell.h"
#import "RecommendedItem.h"

@implementation RecommendedItemsCell

static NSString *const reuseIdentifier = @"ItemCell";

#pragma mark <UICollectionViewDataSource>

- (NSInteger)numberOfSectionsInCollectionView:
    (UICollectionView *)collectionView {
    return 1;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView
     numberOfItemsInSection:(NSInteger)section {
    return self.items.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView
                  cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    RecommendedItem *cell =
        [collectionView dequeueReusableCellWithReuseIdentifier:reuseIdentifier
                                                  forIndexPath:indexPath];

    // Configure the cell

    NSAssert(cell, @"No cell");

    Item *item = self.items[indexPath.row];

    [cell.title setText:item.title];
    cell.item = item;
    NSURL *url = [NSURL URLWithString:item.image];

    NSURLSessionTask *task = [[NSURLSession sharedSession]
          dataTaskWithURL:url
        completionHandler:^(NSData *_Nullable data,
                            NSURLResponse *_Nullable response,
                            NSError *_Nullable error) {

          if (!data) {
              return;
          }

          UIImage *image = [UIImage imageWithData:data];
          if (!image) {
              return;
          }

          dispatch_async(dispatch_get_main_queue(), ^{
            RecommendedItem *updateCell =
                (id)[collectionView cellForItemAtIndexPath:indexPath];
            if (!updateCell) {
                return;
            }
            [updateCell.image setImage:image];
          });

        }];
    [task resume];

    return cell;
}

#pragma mark <UICollectionViewDelegate>

@end

File: EmarsysMobileShopSwift/EmarsysMobileShopSwift/RecommendedItemsCell.swift
//
//  RecommendedItemsCell.swift
//  EmarsysMobileShopSwift
//

import UIKit

class RecommendedItemsCell: UITableViewCell, UICollectionViewDelegate, UICollectionViewDataSource {
    
    @IBOutlet weak var recommendedCollectionView: UICollectionView!
    
    var items = [Item]()
    
    func numberOfSectionsInCollectionView(collectionView: UICollectionView) -> Int {
        return 1
    }
    
    func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return items.count
    }
    
    func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCellWithReuseIdentifier("ItemCell", forIndexPath:indexPath) as! RecommendedItem
        
        // Configure the cell
        let item = items[indexPath.row]
        
        cell.title.text = item.title
        cell.item = item
        
        let url = NSURL(string: item.image!)
        let session = NSURLSession.sharedSession()
        session.dataTaskWithURL(url!, completionHandler: {(data, response, error) in
            guard let data = data else { return }
            guard let image = UIImage(data: data) else { return }
            dispatch_async(dispatch_get_main_queue()) {
                guard let cell = collectionView.cellForItemAtIndexPath(indexPath) else {
                    NSLog("Failed to get cell for index path %@", indexPath)
                    return
                }
                let updateCell = cell as! RecommendedItem
                updateCell.image.image = image
            }
        }).resume()
        
        return cell
    }
}
